File: 17-letter-combinations-of-a-phone-number/letter-combinations-of-a-phone-number.java
class Solution {
    private List<String> combinations = new ArrayList<>();
    private Map<Character, String> letters = Map.of(
        '2',
        "abc",
        '3',
        "def",
        '4',
        "ghi",
        '5',
        "jkl",
        '6',
        "mno",
        '7',
        "pqrs",
        '8',
        "tuv",
        '9',
        "wxyz"
    );
    private String phoneDigits;
    public List<String> letterCombinations(String digits) {
        if(digits.length() == 0) return combinations;

        phoneDigits = digits;
        backtrack(0, new StringBuilder());
        return combinations;
    }

    private void backtrack(int i, StringBuilder str) {
        if (str.length() == phoneDigits.length()) {
            combinations.add(str.toString());
            return;
        }

        String possLet = letters.get(phoneDigits.charAt(i));
        for(char letter : possLet.toCharArray()) {
            str.append(letter);
            backtrack(i+1, str);
            str.deleteCharAt(str.length() - 1);
        }
    }
}

File: 17-letter-combinations-of-a-phone-number/letter-combinations-of-a-phone-number.py
class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        mapping = {
            "2" : ["a", "b", "c"],
            "3" : ["d", "e", "f"],
            "4" : ["g", "h", "i"],
            "5" : ["j", "k", "l"],
            "6" : ["m", "n", "o"],
            "7" : ["p", "q", "r", "s"],
            "8" : ["t", "u", "v"],
            "9" : ["w", "x", "y", "z"]
        }

        ans = []

        def bt(curr, idx):
            if idx == len(digits):
                ans.append(curr)
                return

            dig = digits[idx]
            for char in mapping[dig]:
                bt(curr + char, idx + 1)
        
        bt("", 0)
        return ans